File: src/main/java/nz/co/searchwellington/controllers/permissions/EditPermissionService.java
package nz.co.searchwellington.controllers.permissions;

import com.google.common.base.Strings;
import nz.co.searchwellington.controllers.LoggedInUserFilter;
import nz.co.searchwellington.model.*;
import nz.co.searchwellington.model.frontend.FrontendResource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class EditPermissionService {
	
	private LoggedInUserFilter loggedInUserFilter;
	
	@Autowired
	public EditPermissionService(LoggedInUserFilter loggedInUserFilter) {	
		this.loggedInUserFilter = loggedInUserFilter;
	}
	
	public boolean canEdit(FrontendResource resource) {
        return isAdminOrOwner(resource, loggedInUserFilter.getLoggedInUser());
	}
	
	public boolean canEdit(Resource resource) {
        return isAdminOrOwner(resource, loggedInUserFilter.getLoggedInUser());
	}
	
	public boolean canDelete(FrontendResource resource) {
        return isAdminOrOwner(resource, loggedInUserFilter.getLoggedInUser());
	}
	
	public boolean canDelete(Resource resource) {
        return isAdminOrOwner(resource, loggedInUserFilter.getLoggedInUser());
	}
	
	public boolean canAcceptAll() {
        return isAdmin(loggedInUserFilter.getLoggedInUser());
	}

	public boolean canAcceptAllFrom(Feed feed) {
        return canAcceptAll();
	}
	
	public boolean canCheck(FrontendResource resource) {
        return isAdmin(loggedInUserFilter.getLoggedInUser());
	}

	public boolean canSeeLocalPage(Newsitem newsitem) {
        return isAdmin(loggedInUserFilter.getLoggedInUser());
	}

	public boolean canEditSuggestions() {
		User loggedInUser = loggedInUserFilter.getLoggedInUser();
		return isAdmin(loggedInUser);
	}

	public boolean canAddTag() {
        return isAdmin(loggedInUserFilter.getLoggedInUser());
	}
	
	public boolean canEdit(Tag tag) {
		User loggedInUser = loggedInUserFilter.getLoggedInUser();		
		return loggedInUser != null && loggedInUser.isAdmin();
	}
	
	public boolean canAddWatchlistAndTag() {
        return isAdmin(loggedInUserFilter.getLoggedInUser());
	}

    public boolean canAcceptFeedItems(User loggedInUser) {
        return isAdmin(loggedInUser);
    }

    public boolean canDeleteTags(User loggedInUser) {
        return isAdmin(loggedInUser);
    }
	
	private boolean isAdminOrOwner(FrontendResource resource, User loggedInUser) {
        if (isAdmin(loggedInUser)) {
            return true;
        }
		if (!Strings.isNullOrEmpty(resource.getOwner()) && loggedInUser.getProfilename() == resource.getOwner()) {
			return true;
		}
		return false;
	}

    private boolean isAdmin(User loggedInUser) {
        if (loggedInUser == null) {
            return false;
        }
        return loggedInUser.isAdmin();
    }

	private boolean isAdminOrOwner(Resource resource, User loggedInUser) {
		if (loggedInUser == null) {
			return false;
		}
		if (loggedInUser.isAdmin()) {
			return true;
		}	
		if (resource.getOwner() != null && loggedInUser.getId() == resource.getOwner().getId()) {
			return true;
		}
		return false;
	}

}

File: src/main/scala/nz/co/searchwellington/permissions/EditPermissionService.scala
package nz.co.searchwellington.permissions

import com.google.common.base.Strings
import nz.co.searchwellington.controllers.LoggedInUserFilter
import nz.co.searchwellington.model._
import nz.co.searchwellington.model.frontend.FrontendResource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class EditPermissionService @Autowired()() {

  def canEdit(resource: Resource, loggedInUser: Option[User]): Boolean = {
    isAdminOrOwner(resource, loggedInUser)
  }

  def canDelete(resource: FrontendResource, loggedInUser: Option[User]): Boolean = {
    isAdminOrOwner(resource, loggedInUser)
  }

  def canDelete(resource: Resource, loggedInUser: Option[User]): Boolean = {
    isAdminOrOwner(resource, loggedInUser)
  }

  def canAcceptAll(loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canAcceptAllFrom(feed: Feed, loggedInUser: Option[User]): Boolean = {
    canAcceptAll(loggedInUser)
  }

  def canCheck(resource: FrontendResource, loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canSeeLocalPage(newsitem: Newsitem, loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canEditSuggestions(loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canAddTag(loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canEdit(tag: Tag, loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canAddWatchlistAndTag(loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  def canAcceptFeedItems(loggedInUser: User): Boolean = {
    isAdmin(Some(loggedInUser))
  }

  def canDeleteTags(loggedInUser: Option[User]): Boolean = {
    isAdmin(loggedInUser)
  }

  private def isAdmin(loggedInUser: Option[User]): Boolean = {
    loggedInUser.exists(_.isAdmin)
  }

  private def isAdminOrOwner(resource: FrontendResource, loggedInUser: Option[User]): Boolean = {
    if (isAdmin(loggedInUser)) {
      return true
    }
    loggedInUser.exists { u =>
      val matchesOwnersName = !Strings.isNullOrEmpty(resource.getOwner) && u.getProfilename == resource.getOwner  // TODO proper _id match please
      matchesOwnersName
    }
  }

  private def isAdminOrOwner(resource: Resource, loggedInUser: Option[User]): Boolean = {
    def ownedByLoggedInUser: Boolean = loggedInUser.exists { u =>
      resource.owner.contains(u._id)
    }
    isAdmin(loggedInUser) || ownedByLoggedInUser
  }

}